packages/material-ui-lab/src/Masonry/Masonry.js

```diff
@@ -0,0 +1,142 @@
+// A masonry component using the following libs as inspiration.
+//
+// For the implementation:
+// - https://github.com/STRML/react-grid-layout
+// - https://material.angularjs.org/latest/demo/gridList
+// - https://next.material-ui.com/components/image-list/#masonry-image-list
+// - https://github.com/desandro/masonry
+// - https://mdbootstrap.com/docs/react/layout/masonry/
+// - https://muuri.dev/
+// - https://github.com/bigbite/macy.js
+// - https://github.com/paulcollett/react-masonry-css
+// - https://bvaughn.github.io/react-virtualized/#/components/Masonry
+// - https://github.com/jaredLunde/masonic
+
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import { unstable_composeClasses as composeClasses } from '@material-ui/unstyled';
+import { styled, useThemeProps } from '@material-ui/core/styles';
+import { getMasonryUtilityClass } from './masonryClasses';
+
+const useUtilityClasses = (styleProps) => {
+  const { classes } = styleProps;
+
+  const slots = {
+    root: ['root'],
+  };
+
+  return composeClasses(slots, getMasonryUtilityClass, classes);
+};
+
+const MasonryRoot = styled('div', {
+  name: 'MuiMasonry',
+  slot: 'Root',
+  overridesResolver: (props, styles) => {
+    return [styles.root];
+  },
+})(() => {
+  /* Styles applied to the root element. */
+  return {
+    display: 'grid',
+    gridAutoRows: 0,
+    padding: 0,
+    width: '100%',
+  };
+});
+
+const Masonry = React.forwardRef(function Masonry(inProps, ref) {
+  const props = useThemeProps({
+    props: inProps,
+    name: 'MuiMasonry',
+  });
+
+  const {
+    children,
+    className,
+    cols = 4,
+    component = 'div',
+    gap = 10,
+    style: styleProp,
+    ...other
+  } = props;
+  const [masonryItems, setMasonryItems] = React.useState(children);
+  const style = { gridTemplateColumns: `repeat(${cols}, 1fr)`, gap, ...styleProp };
+  const styleProps = { ...props, component, gap };
+  const classes = useUtilityClasses(styleProps);
+  const resizeItems = () => {
+    const resizedChildren = React.Children.map(children, (child) => {
+      return React.cloneElement(child, {
```

Comment:
Let's try to avoid cloning elements. We can propagate the `gap` either via prop/context, do the calculation inside the `MasonaryItem` and add the `gridRowEnd` via the `styled()` util.